lib/events/athena/consumer.go

```diff
@@ -0,0 +1,439 @@
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package athena
+
+import (
+	"context"
+	"encoding/base64"
+	"errors"
+	"io"
+	"net/url"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/feature/s3/manager"
+	"github.com/aws/aws-sdk-go-v2/service/s3"
+	s3Types "github.com/aws/aws-sdk-go-v2/service/s3/types"
+	"github.com/aws/aws-sdk-go-v2/service/sqs"
+	sqsTypes "github.com/aws/aws-sdk-go-v2/service/sqs/types"
+	"github.com/gravitational/trace"
+	log "github.com/sirupsen/logrus"
+	"golang.org/x/exp/slices"
+
+	apievents "github.com/gravitational/teleport/api/types/events"
+	"github.com/gravitational/teleport/lib/backend"
+	"github.com/gravitational/teleport/lib/events"
+	"github.com/gravitational/teleport/lib/utils"
+)
+
+const (
+	// maxWaitTimeOnReceiveMessageFromSQS defines how long single
+	// receiveFromQueue will wait if there is no max events (10).
+	maxWaitTimeOnReceiveMessageFromSQS = 5 * time.Second
+	// maxNumberOfWorkers defines how many workers are processing messages
+	// from queue or writing parquet files to s3.
+	maxNumberOfWorkers = 5
+)
+
+// consumer is responsible for receiving messages from SQS, batching them up to
+// certain size or interval, and writes to s3 as parquet file.
+type consumer struct {
+	*log.Entry
+	sqsCli              sqsCli
+	backend             backend.Backend
+	storeLocationPrefix string
+	storeLocationBucket string
+	batchMaxItems       int
+	batchMaxInterval    time.Duration
+
+	collectConfig sqsCollectConfig
+}
+
+type sqsCollectConfig struct {
+	sqsCli            sqsCli
+	queueURL          string
+	payloadBucket     string
+	payloadDownloader s3downloader
+	// visibilityTimeout defines how long message won't be available for other
+	// receiveMessage calls. If timeout happens, and message was not deleted
+	// it will return to the queue.
+	visibilityTimeout     int32
+	waitOnReceiveTimeout  int32
+	waitOnReceiveDuration time.Duration
+
+	batchMaxItems int
+}
+
+type sqsCli interface {
+	ReceiveMessage(ctx context.Context, params *sqs.ReceiveMessageInput, optFns ...func(*sqs.Options)) (*sqs.ReceiveMessageOutput, error)
+}
+
+type s3downloader interface {
+	Download(ctx context.Context, w io.WriterAt, input *s3.GetObjectInput, options ...func(*manager.Downloader)) (n int64, err error)
+}
+
+func newConsumer(cfg Config, awsCfg aws.Config, backend backend.Backend, logEntry *log.Entry) (*consumer, error) {
+	u, err := url.Parse(cfg.LocationS3)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	largeEventsURL, err := url.Parse(cfg.LargeEventsS3)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
```

Comment:
yes, I added comment to myself to move it later. `largeEventsURL` was already done in other PR which is also under review. I will rebase it with publisher when it's merge so it will be done.

Reply:
done in https://github.com/gravitational/teleport/pull/24038/commits/d47ce849ccf48838fdf19faea905366369ee7327